ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java

```diff
@@ -266,14 +269,19 @@ private Expression getHavingExpressions() {
         schemaKTable.getKeySerde(),
         SchemaKStream.Type.AGGREGATE,
         ksqlConfig,
-        functionRegistry
+        functionRegistry,
+        aggregationContext.getQueryContext()
     );
 
     if (getHavingExpressions() != null) {
-      result = result.filter(getHavingExpressions());
+      result = result.filter(
+          getHavingExpressions(),
+          contextBuilder.push(FILTER_OP_NAME));
```

Comment:
Is this `FILTER_OP_NAME` being added twice? It's added once here and also again in `SchemaKStream#filter()`.

Relatedly, it feels a little strange that the contexts are being updated in two places: `PlanNode#buildStream()` and in individual methods of `SchemaKStream`. It would be more consistent to remove the calls to `ContextBuilder#push()` from the `SchemaKStream` methods and instead add them to `FilterNode#buildStream()` and `ProjectNode#buildStream()`.

Reply:
They are different levels of context. Here we're pushing context about what the PlanNode is doing (preparing by applying the initial projection, grouping, applying the aggregation, doing the final projection, and applying the where clause). The context being pushed in SchemaKStream is about what's happening at that level, e.g. to distinguish from other things that may be happening in the parent context (e.g. serializing/deserializing).